### ISSUE TITLE
Testing code for FAN model about NYT dataset

### ISSUE THREAD

**Issue by XinZhao0211** [nju-websoft/FAN#2]
Do you have **testing code** about F1 and AUC metrics， i wanna evaluate this model in NYT dataset after get FAN.pkl
thank!

**Comment by koala-222** [comment]
The evaluate code is below:

~~~python
def evaluate(model, data_loader, config):
    avg_loss = AverageMeter()
    logits_history, bag_labels_history = [], []
    with torch.no_grad():
        model.eval()
        bar = tqdm(data_loader)
        for raw_batch in bar:
            inputs, labels = get_inputs(raw_batch, config)
            logits = model(**inputs)
            loss = model.get_loss(logits, labels)

            logits_history.append(deal_with_logits(logits["logits"], config.deal_with_logits, config))  # threshold
            bag_labels_history.append(labels["bag_labels"].cpu().numpy())

            avg_loss.update(loss.item(), 1)
            bar.set_postfix(loss=avg_loss.avg)

    y_ = np.concatenate(logits_history, axis=0).astype(np.float)[:, 1:].reshape(-1)
    y = np.concatenate(bag_labels_history, axis=0).astype(np.int)[:, 1:].reshape(-1)
    assert len(y) == len(y_)
    precision, recall, threshold = metrics.precision_recall_curve(y, y_)
    np.save("precision.npy", precision)
    np.save("recall.npy", recall)
    auc = metrics.auc(recall, precision)
    f1 = (2 * precision * recall / (precision + recall + 1e-20)).max()
    mean_prec = precision.mean()

    logger.info('AUC: {:.4f}'.format(auc))
    logger.info('Micro F1: {:.4f}'.format(f1))
    p100 = P_at_N(precision, 100)
    p200 = P_at_N(precision, 200)
    p300 = P_at_N(precision, 300)
    p500 = P_at_N(precision, 500)
    logger.info("P@100: {:.4f}".format(p100))
    logger.info("P@200: {:.4f}".format(p200))
    logger.info("P@300: {:.4f}".format(p300))
    logger.info("P@500: {:.4f}".format(p500))
    return {'micro_p': precision, 'micro_r': recall, 'micro_p_mean': mean_prec, 'micro_f1': f1, 'auc': auc,
            "p100": p100, "p200": p200, "p300": p300, "p500": p500}


class AverageMeter(object):
    """
    Computes and stores the average and current value of metrics.
    """

    def __init__(self):
        self.reset()

    def reset(self):
        self.val = 0
        self.avg = 0
        self.sum = 0
        self.count = 0

    def update(self, val, n=0):
        self.val = val
        self.sum += val * n
        self.count += n
        self.avg = self.sum / (.0001 + self.count)

    def __str__(self):
        """
        String representation for logging
        """
        # for values that should be recorded exactly e.g. iteration number
        if self.count == 0:
            return str(self.val)
        # for stats
        return '%.4f (%.4f)' % (self.val, self.avg)


def deal_with_logits(logits: torch.Tensor, method, config):
    """
    using max or entropy trick for better handling N/A type
    :param logits: (bags, class_num)
    :param method:
    :param config:
    :return:
    """
    if method == "none":
        return logits.cpu().numpy()
    elif method == "max":
        max_threshold = config.max_threshold
        logits = logits.cpu().numpy()
        flag = (np.max(logits, axis=1) > max_threshold)  # (bags, ): bool
        logits = logits * flag[:, np.newaxis]  # set below max_threshold to 0
        return logits
    elif method == "entropy":
        entropy_threshold = config.entropy_threshold
        logits = logits.cpu().numpy()
        entropy = -np.sum(logits * np.log(logits + 1e-8), axis=1)
        flag = (entropy < entropy_threshold)
        logits = logits * flag[:, np.newaxis]  # set above entropy_threshold to 0
        return logits
    else:
        raise NotImplementedError


def P_at_N(precision, k):
    """compute Precision@N automatically as in OpenNRE
    referring:
        https://github.com/thunlp/OpenNRE/issues/127
        https://github.com/thunlp/OpenNRE/issues/21
        https://github.com/thunlp/OpenNRE/issues/278
    :param precision: np.array()
    :param k:
    :return
    """
    return precision[-k]


~~~
